projects/client/RabbitMQ.Client/src/client/impl/Frame.cs

```diff
@@ -1,218 +1,327 @@
+//  Version 1.1 (the "License"); you may not use this file except in
+//  compliance with the License. You may obtain a copy of the License
+//  at http://www.mozilla.org/MPL/
+//
+//  Software distributed under the License is distributed on an "AS IS"
+//  basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+//  the License for the specific language governing rights and
+//  limitations under the License.
+//
+//  The Original Code is RabbitMQ.
+//
+//  The Initial Developer of the Original Code is Pivotal Software, Inc.
+//  Copyright (c) 2007-2016 Pivotal Software, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+using RabbitMQ.Client.Exceptions;
+using RabbitMQ.Client.Framing;
+using RabbitMQ.Util;
+using System;
+using System.IO;
+
+#if NETFX_CORE
+using Windows.Networking.Sockets;
+#else
+
+using System.Net.Sockets;
+
+#endif
+
+namespace RabbitMQ.Client.Impl
+{
+    public class HeaderWriteFrame : WriteFrame 
+    {
+        public HeaderWriteFrame(int channel, ContentHeaderBase header, int bodyLength) : base(FrameType.FrameHeader, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            
+            writer.Write((ushort)header.ProtocolClassId);
+            header.WriteTo(writer, (ulong)bodyLength);
+        }
+    }
+    
+    public class BodySegmentWriteFrame : WriteFrame
+    {
+        public BodySegmentWriteFrame(int channel, byte[] body, int offset, int count) : base(FrameType.FrameBody, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            
+            writer.Write(body, offset, count);
+        }
+    }
+    
+    public class MethodWriteFrame : WriteFrame
+    {
+        public MethodWriteFrame(int channel, MethodBase method) : base(FrameType.FrameMethod, channel)
+        {
+            NetworkBinaryWriter writer = base.GetWriter();
+            
+            writer.Write((ushort)method.ProtocolClassId);
+            writer.Write((ushort)method.ProtocolMethodId);
+            
+            var argWriter = new MethodArgumentWriter(writer);
+            
+            method.WriteArgumentsTo(argWriter);
+            
+            argWriter.Flush();
+        }
+    }
+    
+    public class EmptyWriteFrame : WriteFrame
+    {
+        private static readonly byte[] m_emptyByteArray = new byte[0];
+
+        public EmptyWriteFrame() : base(FrameType.FrameHeartbeat, 0)
+        {
+            base.GetWriter().Write(m_emptyByteArray);
+        }
+        
+        public override string ToString()
+        {
+            return base.ToString() + string.Format("(type={0}, channel={1}, {2} bytes of payload)",
+                Type,
+                Channel,
+                Payload == null
+                    ? "(null)"
+                    : Payload.Length.ToString());
+        }
+    }
+    
+    public class WriteFrame : Frame

```

Comment:
Can we please rename this to `WriteableFrame` or `OutgoingFrame`? There are no "write frames" in the protocol (that would be different from "read frames") plus we have methods that are named `WriteFrame` and it gets fairly confusing even with a decent IDE.